File: 2025-05-MAY-GFG-POTD/May 10 - Longest Subarray with Majority Greater than K/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int longestSubarray(vector<int>& arr, int k) {
        int n = arr.size();
        vector<int> temp(n);

        // Convert array elements
        for (int i = 0; i < n; ++i) {
            temp[i] = (arr[i] > k) ? 1 : -1;
        }

        unordered_map<int, int> prefixMap;
        int prefixSum = 0, maxLength = 0;

        for (int i = 0; i < n; ++i) {
            prefixSum += temp[i];

            if (prefixSum > 0) {
                maxLength = i + 1;
            } else {
                if (prefixMap.find(prefixSum) == prefixMap.end()) {
                    prefixMap[prefixSum] = i;
                }
                if (prefixMap.find(prefixSum - 1) != prefixMap.end()) {
                    maxLength = max(maxLength, i - prefixMap[prefixSum - 1]);
                }
            }
        }

        return maxLength;
    }
};

// Driver code
int main() {
    int t;
    cin >> t;
    cin.ignore();

    while (t--) {
        string line;
        getline(cin, line);
        stringstream ss(line);
        vector<int> arr;
        int x;
        while (ss >> x) {
            arr.push_back(x);
        }

        int k;
        cin >> k;
        cin.ignore();

        Solution sol;
        cout << sol.longestSubarray(arr, k) << endl;
        cout << "~" << endl;
    }
    return 0;
}

File: 2025-05-MAY-GFG-POTD/May 10 - Longest Subarray with Majority Greater than K/Solution.py
class Solution:
    def longestSubarray(self, arr, k):
        n = len(arr)
        temp = [1 if x > k else -1 for x in arr]

        prefix_sum = 0
        max_length = 0
        prefix_map = {}

        for i in range(n):
            prefix_sum += temp[i]

            if prefix_sum > 0:
                max_length = i + 1
            else:
                if prefix_sum not in prefix_map:
                    prefix_map[prefix_sum] = i

                if (prefix_sum - 1) in prefix_map:
                    max_length = max(max_length, i - prefix_map[prefix_sum - 1])

        return max_length

# Driver code
if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        arr = list(map(int, input().strip().split()))
        k = int(input())
        sol = Solution()
        print(sol.longestSubarray(arr, k))
        print("~")
